### PR TITLE
[#175896908] add get cgn status api

### PR DESCRIPTION
**Pull request by AleDore**
#### List of Changes
- Add `GetCgnStatus` API
- Configure new CGN endpoint for io-functions-cgn
- Add API Client for CGN
- Add CGN's openapi specs

#### Motivation and Context
In order to provide an API that can return information about CGN status, we have to expose an API to IO-APP that can be also used to obtain additional info related to a specific CGN status. See the related Pivotal [story](https://www.pivotaltracker.com/story/show/175896908) for further informations.

#### How Has This Been Tested?
It has been tested by performing unit tests.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

P.S: We have to create a related PR into the [infrastructure](https://github.com/pagopa/io-infrastructure-live-new) repo to align Env variables properly.
#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -36,7 +36,9 @@
     "generate:proxy:public-models": "rimraf generated/public  && gen-api-models --api-spec api_public.yaml --out-dir generated/public",
     "generate:api:io": "rimraf generated/io-api && gen-api-models --api-spec https://raw.githubusercontent.com/pagopa/io-functions-app/master/openapi/index.yaml --no-strict --out-dir generated/io-api --request-types --response-decoders --client",
     "generate:proxy:bonus-models": "rimraf generated/bonus && gen-api-models --api-spec api_bonus.yaml --out-dir generated/bonus",
+    "generate:proxy:cgn-models": "rimraf generated/cgn && gen-api-models --api-spec api_cgn.yaml --out-dir generated/cgn",
     "generate:api:io-bonus": "rimraf generated/io-bonus-api && gen-api-models --api-spec https://raw.githubusercontent.com/pagopa/io-functions-bonus/master/openapi/index.yaml --no-strict --out-dir generated/io-bonus-api --request-types --response-decoders --client",
+    "generate:api:io-cgn": "rimraf generated/io-cgn-api && gen-api-models --api-spec https://raw.githubusercontent.com/pagopa/io-functions-cgn/175896908_add_get_cgn_status_api/openapi/index.yaml --no-strict --out-dir generated/io-cgn-api --request-types --response-decoders --client",
```
Comment (balanza) [inline package.json:41]: Should this be master?

Comment (AleDore) [reply package.json:41]: Yep :) I will update when we merge this [https://github.com/pagopa/io-functions-cgn/pull/5](https://github.com/pagopa/io-functions-cgn/pull/5)

**Review by gunzip — APPROVED** [review]

File: .env.example [hunk 2]
```diff
@@ -43,8 +43,13 @@ USERS_LOGIN_QUEUE_NAME=ulqueuename
...
+FF_CGN_ENABLED=1
 JWT_SUPPORT_TOKEN_PRIVATE_RSA_KEY="-----BEGIN RSA PRIVATE KEY-----\n
 <APrivateRSAKEY>\n
 -----END RSA PRIVATE KEY-----"
 JWT_SUPPORT_TOKEN_ISSUER=io-backend
 SPID_LEVEL_WHITELIST=SpidL2,SpidL3
+
+CGN_API_KEY=put_your_api_key_here
+CGN_API_URL=http://host.docker.internal:7073/api/v1
+CGN_API_BASE_PATH="/cgn/api/v1"
```
Comment (gunzip) [inline .env.example:55]: can you put these into io-mock as well?

File: README.md [hunk 3]
```diff
@@ -11,7 +11,7 @@ This repository contains the code of the backend used by the
 
 ## Table of content
 
-- [IO mobile backend](#io-mobile-backend)
+- [Digital citizenship web and mobile backend](#digital-citizenship-web-and-mobile-backend)
```
Comment (gunzip) [inline README.md:14]: why this regression? (maybe need rebasing)

**Review by gunzip — DISMISSED** [review]

File: .env.example (hunk 2 above)
Comment (AleDore) [reply .env.example:55]: Yup

File: README.md (hunk 3 above)
Comment (AleDore) [reply README.md:14]: Rebased :)

**Comment by AleDore** [comment]
Can we merge this? @balanza @gunzip

File: README.md (hunk 3 above)
Comment (balanza) [reply README.md:14]: I this to be resolved yet?

File: README.md [hunk 4]
```diff
@@ -39,8 +39,8 @@ This repository contains the code of the backend used by the
     - [SPID Identity Providers Metadata](#spid-identity-providers-metadata)
     - [SPID Validator](#spid-validator)
     - [Architecture decision records](#architecture-decision-records)
+  - [| 6   | Backend is deployed on more than one instance |                    |](#-6----backend-is-deployed-on-more-than-one-instance---------------------)
```
Comment (balanza) [inline README.md:42]: it's seems like this line is in the wrong position

File: README.md [hunk 5]
```diff
@@ -146,6 +146,10 @@ Those are all Environment variables needed by the application:
 | BONUS_API_KEY                          | The key used to authenticate to the io-functions-bonus API                        | string |
 | BONUS_API_URL                          | The io-functions-bonus  URL                                                       | string |
 | BONUS_API_BASE_PATH                    | The root path for the backend bonus api endpoints                                 | string |
+| CGN_API_KEY                            | The key used to authenticate to the io-functions-cgn API                          | string |
+| CGN_API_URL                            | The io-functions-cgn  URL                                                         | string |
+| CGN_API_BASE_PATH                      | The root path for the backend cgn api endpoints                                   | string |
+| BONUS_REQUEST_LIMIT_DATE               | The date until a user can request a bonus, in UTC ISO format                      | string |
```
Comment (balanza) [inline README.md:152]: We removed this variable, why has it appeared again?

**Review by balanza — CHANGES_REQUESTED** [review]
It seems ok despite some glitch from the rebase

**Comment by AleDore** [comment]
Can we move on on this? @balanza @BurnedMarshal

File: src/services/__tests__/cgnService.test.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+      });
+      const service = new CgnService(api);
+  
+      const res = await service.getCgnStatus(mockedUser);
+  
+      expect(res).toMatchObject({
+        kind: "IResponseErrorInternal"
+      });
+    });
+  });
```
Comment (BurnedMarshal) [inline src/services/__tests__/cgnService.test.ts:138]: newline EOF

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]
More changes were made inside another PR #773

File: src/app.ts [hunk 7]
```diff
@@ -809,6 +827,22 @@ function registerSessionAPIRoutes(
...
+  app: Express,
+  basePath: string,
+  cgnService: CgnService,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  bearerSessionTokenAuth: any
+): void {
+  const cgnController: CgnController = new CgnController(cgnService);
+
+  app.get(
+    `${basePath}/cgn/status`,
```
[the remark this answers is not in the document]
Comment (AleDore) [reply src/app.ts:840]: Yes

File: src/__tests__/app.test.ts [hunk 8]
```diff
@@ -63,6 +63,7 @@ const aBonusAPIBasePath = "/bonus/api/v1";
 const aPagoPABasePath = "/pagopa/api/v1";
 const aMyPortalBasePath = "/myportal/api/v1";
 const aBPDBasePath = "/bpd/api/v1";
+const aCgnAPIBasePath = "/cgn/api/v1";
```
[the remark this answers is not in the document]
Comment (AleDore) [reply src/__tests__/app.test.ts:66]: It has been copied from BonusAPIBasePath :) and it has the same problem. Anyway the base path should be `/api/v1/cgn`

**Review by BurnedMarshal — DISMISSED** [review]
lgtm

**Comment by balanza** [comment]
LGTM besides the followings:
* https://github.com/pagopa/io-backend/pull/751/files#r572991545
* https://github.com/pagopa/io-backend/pull/751/files#r561883015

**Comment by AleDore** [comment]
> LGTM besides the followings:
> 
> * https://github.com/pagopa/io-backend/pull/751/files#r572991545
> * https://github.com/pagopa/io-backend/pull/751/files#r561883015

fixed [here](https://github.com/pagopa/io-backend/pull/751/commits/b6020a772869e1024b5d66cf4c4198898b250d78)

**Review by balanza — APPROVED** [review]

**Pull request merged**